realtek: rt-loader: allow arbitrary kernel load address

jonasjelonek opened the pull request:
This allows me to RAM-boot on Zyxel XS1930 devices. Using the default setup where   
```
<kernel load address> == <rt-loader initial run address> == 0x80100000
```
doesn't work because BootExt presumably uses some RAM in that area. Switching to `0x80300000` (based on the memory map) works fine but would require to change the kernel load address. This cannot just be done with changing `KERNEL_LOADADDR` but would require e.g. changing `arch/mips/rtl83xx/Platform` (and probably more).

Log with this change:
```
XS1930-12F> ATUP80300000,5133213
...
OK
XS1930-12F> ATGO80300000
rt-loader
Running on RTL9313 (chip id 6567A) with 256MB
Relocate 5133216 bytes from 0x80300000 to 0x8fac0000
Extract image with 5117505 bytes from 0x8fac3d5c to 0x80100000 ...
Final kernel size is 19602473 bytes
Booting kernel from 0x80100000 ...

<kernel boots>
```

This is probably only needed for RAM-boot/Initramfs image. Loading from flash, which is needed here too for full support, can use the standalone variant of rt-loader.

jonasjelonek committed:
realtek: rt-loader: allow arbitrary kernel load address

rt-loader currently has two operation modes, piggy-backed and
standalone. In standalone mode, the kernel load address is read from the
uImage in flash. In piggy-backed mode, rt-loader instead uses its
initial run address (aka run address during first run) as the kernel
load address. This is safe and works fine for all devices either using
U-boot or having no issue uploading an image to the default kernel load
address 0x80100000.

To extend usecases, allow to specify a kernel load address when
building rt-loader. In this case, rt-loader uses this address instead of
the address inferred at runtime.

On certain Zyxel devices, this allows to upload and boot an rt-loader
piggy-backed image to an alternate address but keep the default kernel
load address of 0x80100000. BootExt on these devices occupies memory
above and will crash during transfer when this address is used as upload
location. Using this extension, the image can be uploaded to e.g.
0x80300000 and rt-loader will use 0x80100000 as the final load address.
This avoid taking the pain the adjust the load address of the kernel
itself.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21248
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul reviewed (changes requested):
You are opening up a new usecase. Good to see.

Why didn't we opt for this? Until now the output format of any build is usually:

- uimage hader (contains load address 0x80100000)
- rt-loader
- lzma compressed kernel

So rt-loader can safely assume that its load address is the target kernel address and was previously calculated by the bootloader from the uimage. In your case you execute the loader from an arbitrary address.

In the flash case https://github.com/openwrt/openwrt/pull/20439 rt-loader already sets the kernel load address to an appropriate value.

jonasjelonek committed:
realtek: image: add rt-loader-no-uimage recipe

Add a new recipe 'rt-loader-no-uimage' that passes the kernel load
address to rt-loader, causing it to use that instead of it's initial run
address.

The usual behavior is fine for uImages where the load address is
predefined in the header, U-boot loads the image to that address and
then runs it, rt-loader just takes over that address. For non-uImage
instead, the address is tightly coupled to where the image has been
transferred during serial or TFTP upload. This may not be possible on
several devices. Passing a separate kernel load address to rt-loader
decouples that and avoids taking the pain to change the load address of
the kernel itself.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21248
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul reviewed (approved):
Thanks
